src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -1702,145 +1854,214 @@ final byte getCoder() {
-                    return;
+                    StringUTF16.putCharsSB(value, j, s, i, end);
+                    return value;
                 }
             }
         } else {
-            StringUTF16.putCharsSB(this.value, index, s, off, end);
-        }
-    }
-
-    private void inflateIfNeededFor(String input) {
-        if (COMPACT_STRINGS && (coder != input.coder())) {
-            inflate();
+            StringUTF16.putCharsSB(value, index, s, off, end);
         }
+        return value;
     }
 
-    private void inflateIfNeededFor(AbstractStringBuilder input) {
-        if (COMPACT_STRINGS && (coder != input.getCoder())) {
-            inflate();
+    private static byte[] inflateIfNeededFor(byte[] value, int count, byte coder, byte otherCoder) {
+        if (COMPACT_STRINGS && (coder == LATIN1 && otherCoder == UTF16)) {
+            return inflateToUTF16(value, count);
         }
+        return value;
     }
 
-    private void putStringAt(int index, String str, int off, int end) {
-        inflateIfNeededFor(str);
-        str.getBytes(value, off, index, coder, end - off);
-    }
-
-    private void putStringAt(int index, String str) {
-        inflateIfNeededFor(str);
-        str.getBytes(value, index, coder);
-    }
-
-    private final void appendChars(char[] s, int off, int end) {
-        int count = this.count;
-        if (isLatin1()) {
-            byte[] val = this.value;
-            for (int i = off, j = count; i < end; i++) {
+    /**
+     * {@return the buffer with the substring inserted}
+     * If the substring contains UTF16 characters and the current coder is LATIN1, inflation occurs
+     * into a new buffer and returned; the caller must update the coder to UTF16.
+     * Since the contents are immutable, a simple copy of characters is sufficient.
+     * @param value an existing buffer to insert into
+     * @param coder the coder of the buffer
+     * @param count the count of characters in the buffer
+     * @param index the index to insert the string
+     * @param str the string
+     */
+     private static byte[] putStringAt(byte[] value, byte coder, int count, int index, String str, int off, int end) {
+        byte[] newValue = inflateIfNeededFor(value, count, coder, str.coder());
+        coder = (newValue == value) ? coder : UTF16;
+        str.getBytes(newValue, off, index, coder, end - off);
+        return newValue;
+    }
+
+    /**
+     * {@return buffer with new characters appended, possibly inflated}
+     * The buffer may need to be inflated if any character does not fit;
+     * if the returned buffer is different then passed in, the new coder is UTF16.
+     * The caller is responsible for updating the count.
+     * @param value the current buffer
+     * @param coder the coder of the buffer
+     * @param count the character count
+     * @param s a char array
+     * @param off the offset of the first character to append
+     * @param end end last (exclusive) character to append
```

Comment:
It might be helpful to somehow clarify that, in the case of appending Latin1 characters to a Latin1 buffer, it is expected/required that `value` is already of sufficient size.